File: apps/legacy-cli/modules/infracost.py
import os
import urllib.request
import tarfile
import json
from modules.logs import LOGS
from modules.utils import execute, get_system_architecture

myLOGS = LOGS()

class INFRACOST:
    def __init__(self, version="v0.10.39"):
        self.api_key = os.getenv('INFRACOST_API_KEY')
        self.version = version
        self.binary_path = f"bin/infracost_{self.version}"

    def check_token(self):
        """Check if the Infracost API token is set."""
        if not self.api_key:
            myLOGS.log("warn", "Infracost token not provided. Skipping cost estimation.")
            myLOGS.log("warn", "To include Infracost, run with --infracost-token <YOUR_API_KEY>")
            return False
        return True

    def _download_binary(self):
        """Download and extract the Infracost binary if it doesn't already exist."""
        if os.path.exists(self.binary_path):
            myLOGS.log("debug", f"Infracost {self.version} is already available.")
            return

        system_type, arch = get_system_architecture()
        download_url = f"https://github.com/infracost/infracost/releases/download/{self.version}/infracost-{system_type}-{arch}.tar.gz"
        myLOGS.log("info", f"Downloading Infracost {self.version} for {system_type}-{arch}...")

        os.makedirs("bin", exist_ok=True)
        tar_file_path = f"bin/infracost_{self.version}.tar.gz"
        urllib.request.urlretrieve(download_url, tar_file_path)

        # Extract the first (and expected single) file from the tarball
        with tarfile.open(tar_file_path, "r:gz") as tar:
            extracted_file_name = tar.getnames()[0]
            tar.extractall("bin")

        # Rename the extracted file to the desired binary path
        extracted_file_path = os.path.join("bin", extracted_file_name)
        os.rename(extracted_file_path, self.binary_path)
        os.chmod(self.binary_path, 0o775)
        os.remove(tar_file_path)
        myLOGS.log("info", f"Infracost {self.version} downloaded and extracted successfully.")


    def run_infracost(self, terraform_plan_file="temp/terraform.plan.json"):
        """Run Infracost cost estimation and output JSON and table formats."""
        if not self.check_token():
            return

        self._download_binary()

        # Define file paths for JSON and table output
        breakdown_json_path = "temp/infracost_breakdown.json"
        breakdown_table_path = "temp/infracost_breakdown_table.txt"

        # Run Infracost breakdown command
        myLOGS.log("info", "Running Infracost cost estimation...")
        breakdown_cmd = f"{self.binary_path} breakdown --path {terraform_plan_file} --format json --out-file {breakdown_json_path}"
        try:
            execute(breakdown_cmd, silent=True, log_file="./logs/infracost.breakdown.log")
        except Exception as error:
            myLOGS.log("warning", f"Failed to run Infracost breakdown: {error}. Check the log file ./logs/infracost.breakdown.log")
            return

        # Convert JSON breakdown to a table format
        output_cmd = f"{self.binary_path} output --format table --path {breakdown_json_path} --out-file {breakdown_table_path}"
        try:
            execute(output_cmd)
        except Exception as error:
            myLOGS.log("warning", f"Failed to convert Infracost JSON to table format: {error}")
            return

        # Load and log summary cost information from JSON breakdown
        try:
            with open(breakdown_json_path, "r") as breakdown_file:
                breakdown_data = json.load(breakdown_file)
                total_monthly_cost = breakdown_data.get("totalMonthlyCost", "N/A")
                diff_total_monthly_cost = breakdown_data.get("diffTotalMonthlyCost", "N/A")
                myLOGS.log("info", f"Cost Summary - Total Monthly Cost: ${total_monthly_cost}, Diff Monthly Cost: ${diff_total_monthly_cost}")
        except (json.JSONDecodeError, KeyError) as error:
            myLOGS.log("warning", f"Failed to parse cost summary: {error}")

        # Notify the user about the detailed cost breakdown table
        myLOGS.log("info", f"For detailed cost breakdown, see the table format in {breakdown_table_path}")

File: packages/core/infracost/infracost.go
// SPDX-FileCopyrightText: 2026 Alethia Labs <legal@alethialabs.io>
// SPDX-License-Identifier: AGPL-3.0-only

package infracost

import (
	"archive/tar"
	"compress/gzip"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/alethialabs-io/alethialabs/packages/core/utils"
)

var (
	httpGet        = http.Get
	executeCommand = utils.ExecuteCommand
)

// InfracostCLI represents the Infracost CLI wrapper.
type InfracostCLI struct {
	Version    string
	binaryPath string
	apiKey     string
}

// NewInfracostCLI creates a new Infracost CLI wrapper.
func NewInfracostCLI(version string, apiKey string) *InfracostCLI {
	return &InfracostCLI{
		Version: version,
		apiKey:  apiKey,
	}
}

// CheckToken verifies if the Infracost API token is set.
func (i *InfracostCLI) CheckToken() bool {
	if i.apiKey == "" {
		fmt.Println("Warning: Infracost token not provided. Skipping cost estimation.")
		fmt.Println("To include Infracost, set the INFRACOST_API_KEY environment variable.")
		return false
	}
	return true
}

// ensureBinary checks if the Infracost binary exists and downloads it if not.
func (i *InfracostCLI) ensureBinary() error {
	binDir := "bin"
	absBinDir, err := filepath.Abs(binDir)
	if err != nil {
		return fmt.Errorf("failed to get absolute path for bin directory: %w", err)
	}

	i.binaryPath = filepath.Join(absBinDir, fmt.Sprintf("infracost_%s", i.Version))
	if _, err := os.Stat(i.binaryPath); err == nil {
		fmt.Printf("Infracost %s is already available.\n", i.Version)
		return nil
	}

	// Create bin directory if it doesn't exist
	if err := os.MkdirAll(absBinDir, 0755); err != nil {
		return fmt.Errorf("failed to create bin directory: %w", err)
	}

	return i.download(absBinDir)
}

func (i *InfracostCLI) download(binDir string) error {
	arch := runtime.GOARCH
	goos := runtime.GOOS

	downloadURL := fmt.Sprintf("https://github.com/infracost/infracost/releases/download/%s/infracost-%s-%s.tar.gz", i.Version, goos, arch)
	fmt.Printf("Downloading Infracost %s for %s-%s...\n", i.Version, goos, arch)

	resp, err := httpGet(downloadURL)
	if err != nil {
		return fmt.Errorf("failed to download infracost: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to download infracost: status code %d", resp.StatusCode)
	}

	tarFile := filepath.Join(binDir, fmt.Sprintf("infracost_%s.tar.gz", i.Version))
	out, err := os.Create(tarFile)
	if err != nil {
		return fmt.Errorf("failed to create tar file: %w", err)
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	if err != nil {
		return fmt.Errorf("failed to write tar file: %w", err)
	}

	// Extract the tar.gz file
	f, err := os.Open(tarFile)
	if err != nil {
		return fmt.Errorf("failed to open tar file: %w", err)
	}
	defer f.Close()

	gzReader, err := gzip.NewReader(f)
	if err != nil {
		return fmt.Errorf("failed to create gzip reader: %w", err)
	}
	defer gzReader.Close()

	tarReader := tar.NewReader(gzReader)

	for {
		header, err := tarReader.Next()
		if err == io.EOF {
			break // End of archive
		}
		if err != nil {
			return fmt.Errorf("failed to read tar header: %w", err)
		}

		if header.Typeflag == tar.TypeReg && strings.Contains(header.Name, "infracost") {
			outPath := filepath.Join(binDir, "infracost")
			outFile, err := os.Create(outPath)
			if err != nil {
				return fmt.Errorf("failed to create infracost binary: %w", err)
			}
			defer outFile.Close()

			_, err = io.Copy(outFile, tarReader)
			if err != nil {
				return fmt.Errorf("failed to write infracost binary: %w", err)
			}

			i.binaryPath = outPath // Update binaryPath to the extracted path
			break
		}
	}

	// Make executable
	if err := os.Chmod(i.binaryPath, 0755); err != nil {
		return fmt.Errorf("failed to make infracost binary executable: %w", err)
	}

	// Clean up tar.gz file
	if err := os.Remove(tarFile); err != nil {
		fmt.Printf("Warning: failed to remove tar.gz file %s: %v\n", tarFile, err)
	}

	fmt.Println("Infracost downloaded and extracted successfully.")
	return nil
}

// RunInfracost executes infracost breakdown and returns structured cost data.
// Also writes the JSON and table files to temp/ for backward compatibility.
func (i *InfracostCLI) RunInfracost(planFile string, env []string) (*CostBreakdown, error) {
	if !i.CheckToken() {
		return nil, nil
	}

	fmt.Println("Running Infracost cost estimation...")

	if err := i.ensureBinary(); err != nil {
		return nil, fmt.Errorf("failed to ensure infracost binary: %w", err)
	}

	tempDir := "temp"
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create temp directory: %w", err)
	}
	breakdownJSONPath := filepath.Join(tempDir, "infracost_breakdown.json")

	breakdownCmd := fmt.Sprintf("%s breakdown --path %s --format json --out-file %s", i.binaryPath, planFile, breakdownJSONPath)
	if err := executeCommand(breakdownCmd, ".", env, nil, nil); err != nil {
		return nil, fmt.Errorf("infracost breakdown failed: %w", err)
	}

	data, err := os.ReadFile(breakdownJSONPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read infracost output: %w", err)
	}

	breakdown, err := ParseCostBreakdown(data)
	if err != nil {
		return nil, err
	}

	if breakdown.Summary != nil {
		fmt.Printf("Cost Summary — Monthly: $%.2f, Diff: $%.2f (%d resources with cost, %d free)\n",
			breakdown.Summary.TotalMonthly,
			breakdown.Summary.DiffMonthly,
			breakdown.Summary.ResourcesWithCost,
			breakdown.Summary.ResourcesFree,
		)
	}

	return breakdown, nil
}
